File: Exercises/Exercise 01/Ex01.py

from threading import Thread

j = 0

def thread_1_function():
    global j
    for i in range(0,100000):
        j+=1

def thread_2_function():
    global j
    for i in range(-100000,0):
        j-=1

def main():
    thread_1 = Thread(target = thread_1_function, args = (),)
    thread_1.start()
    thread_2 = Thread(target = thread_2_function, args = (),)
    thread_2.start()

    thread_1.join()
    thread_2.join()
    print(j)

main()

File: Exercises/Exercise 01/Ex01.go
/*
global shared int = 0
main:
  spawn thread_1
  spawn thread_2
  join all threads (wait for them to finish)
  print i
thread_1:
  do 1_000_000 times:
    i++
thread_2:
  do 1_000_000 times:
    i--
*/
package main

import (
	. "fmt"
	"time"
)

var j int

func thread_1() {
	for i := 0; i < 1000000; i++ {
		j++
	}
}

func thread_2() {
	for i := 0; i < 1000000; i++ {
		j--
	}
}

func main() {
	j = 0
	go thread_1()
	go thread_2()
	time.Sleep(1000 * time.Millisecond)
	Println(j)
}